yt_dlp/extractor/jiosaavn.py

```diff
@@ -0,0 +1,98 @@
+from .common import InfoExtractor
+from ..utils import (
+    js_to_json,
+    urlencode_postdata,
+)
+
+
+class JioSaavnBaseInfoExtractor(InfoExtractor):
+    def extract_initial_data_as_json(self, url, id):
+        webpage = self._download_webpage(url, id)
+        init_data_re = r'window.__INITIAL_DATA__\s*=\s*(?P<json>.+?);*\s*\</script>'
+
+        init_data = self._parse_json(self._search_regex(init_data_re, webpage,
+                                                        'init-json'),
+                                     id, transform_source=js_to_json)
+        return init_data
+
+
+class JioSaavnSongIE(JioSaavnBaseInfoExtractor):
+    _VALID_URL = r'''(?x)
+                    https?://(?:www\.)?
+                        (?:
+                            jiosaavn\.com/song/[^/]+/|
+                            saavn.com/s/song/(?:[^/]+/){3}
+                        )
+                        (?P<id>[^/]+)
+                   '''
+    _TESTS = [{
+        'url': 'https://www.jiosaavn.com/song/leja-re/OQsEfQFVUXk',
+        'md5': '7b1f70de088ede3a152ea34aece4df42',
+        'info_dict': {
+            'id': 'OQsEfQFVUXk',
+            'ext': 'mp3',
+            'title': 'Leja Re',
+            'album': 'Leja Re',
+            'thumbnail': 'https://c.saavncdn.com/258/Leja-Re-Hindi-2018-20181124024539-500x500.jpg',
+        },
+    }, {
+        'url': 'https://www.saavn.com/s/song/hindi/Saathiya/O-Humdum-Suniyo-Re/KAMiazoCblU',
+        'only_matching': True,
+    }]
+
+    def _real_extract(self, url):
+        audio_id = self._match_id(url)
+        song_data = self.extract_initial_data_as_json(url, audio_id)['song']['song']
+
+        data = urlencode_postdata({'__call': 'song.generateAuthToken',
+                                   '_format': 'json',
+                                   'bitrate': '128',
+                                   'url': song_data['encrypted_media_url'],
+                                   })
+
+        def clean_api_json(resp):
+            return self._search_regex(r'(?P<json>\{.+?})', resp, 'api-json')
+
+        media_url = self._download_json('https://www.jiosaavn.com/api.php',
+                                        audio_id, data=data,
+                                        transform_source=clean_api_json,
+                                        )['auth_url']
+
+        return {
+            'id': audio_id,
+            'title': song_data['title']['text'],
+            'formats': [{
+                'url': media_url,
+                'ext': 'mp3',
+                'format_note': 'MPEG audio',
+                'format_id': '128',
+                'vcodec': 'none',
+            }],
+            'album': song_data.get('album', {}).get('text'),
+            'thumbnail': song_data.get('image', [None])[0],
```

Comment:
I got returned a mp4 file with AAC audio, so hard-coding mp3 doesn't make sense. We can also make title extraction non-fatal.
I am a bit unsure regarding the `transform_source=clean_api_json`. Is that really needed? The two test cases seem to work without. I have removed it in the suggestion.
```suggestion
        media_data = self._download_json('https://www.jiosaavn.com/api.php', audio_id, data=data)

        return {
            'id': audio_id,
            'formats': [{
                'url': media_data['auth_url'],
                'container': media_data.get('type'),
                'vcodec': 'none',
            }],
            **traverse_obj(song_data, {
                'title': ('title', 'text'),
                'album': ('album', 'text'),
                'thumbnail': ('image', 0, {url_or_none}),
            }),
```
Tests will need to be adjusted and `url_or_none` and `from ..utils.traversal import traverse_obj` added to the imports.